shared/src/ledger/native_vp/governance/utils.rs

```diff
@@ -74,48 +77,315 @@ impl ProposalEvent {
     }
 }
 
-/// Return a proposal result - accepted only when the result is `Ok(true)`.
-pub fn compute_tally<D, H>(
-    storage: &Storage<D, H>,
-    epoch: Epoch,
+/// Return a proposal result
+pub fn compute_tally(
     votes: Votes,
-) -> storage_api::Result<bool>
-where
-    D: DB + for<'iter> DBIter<'iter> + Sync + 'static,
-    H: StorageHasher + Sync + 'static,
-{
-    let total_stake: VotePower = storage.total_stake(epoch)?.into();
-
+    total_stake: VotePower,
+    proposal_type: &ProposalType,
+) -> ProposalResult {
     let Votes {
         yay_validators,
-        yay_delegators,
-        nay_delegators,
+        delegators,
     } = votes;
 
-    let mut total_yay_staked_tokens = VotePower::from(0_u64);
-    for (_, amount) in yay_validators.clone().into_iter() {
-        total_yay_staked_tokens += amount;
-    }
+    match proposal_type {
+        ProposalType::Default(_) | ProposalType::ETHBridge => {
+            let mut total_yay_staked_tokens = VotePower::default();
 
-    // YAY: Add delegator amount whose validator didn't vote / voted nay
-    for (_, vote_map) in yay_delegators.iter() {
-        for (validator_address, vote_power) in vote_map.iter() {
-            if !yay_validators.contains_key(validator_address) {
-                total_yay_staked_tokens += vote_power;
+            for (_, (amount, validator_vote)) in yay_validators.iter() {
+                if let ProposalVote::Yay(vote_type) = validator_vote {
+                    if proposal_type == vote_type {
+                        total_yay_staked_tokens += amount;
+                    } else {
+                        return ProposalResult {
+                            result: TallyResult::Failed(format!(
+                                "Unexpected vote type. Expected: {}, Found: {}",
+                                proposal_type, validator_vote
+                            )),
+                            total_voting_power: total_stake,
+                            total_yay_power: 0,
+                            total_nay_power: 0,
+                        };
+                    }
+                } else {
+                    return ProposalResult {
```

Comment:
Here we return a failed proposal result. The vote type is checked by the governance vp, so a mismatch here is possible only if a bug is present in the vp or someone uses a governance proposal to write the invalid vote in storage. I'm not sure what strategy we should follow here:

1. Panic, since a vote of the wrong type shouldn't be allowed
2. Ignoring the error and continuing, since a vote of the wrong type doesn't prevent us from computing the tally
3. Signaling the error and rejecting the proposal (which is the solution currently implemented)